lib/pwnlib/reg_sort.rb

```diff
@@ -0,0 +1,221 @@
+      #       ['mov', 'x', 'a'],
+      #       ['mov', 'a', 'b'],
+      #       ['mov', 'b', 'c'],
+      #       ['mov', 'c', 'x'],
+      #       ['mov', 'y', 'z']]
+      #   regsort({a: 'b', b: 'c', c: 'a', x: '1', y: 'z', z: 'c'}, regs, xchg: false)
+      #   => [['mov', 'x', '1'],
+      #       ['mov', 'z', 'c'],
+      #       ['mov', 'x', 'a'],
+      #       ['mov', 'a', 'b'],
+      #       ['mov', 'b', 'c'],
+      #       ['mov', 'c', 'x'],
+      #       ['mov', 'y', 'z']]
+      def regsort(in_out, all_regs, tmp: nil, xchg: true, randomize: nil)
+        # randomize = context.randomize if randomize.nil?
+
+        # TODO(david942j): stringify_keys
+        in_out = in_out.map { |k, v| [k.to_s, v] }.to_h
+        # Drop all registers which will be set to themselves.
+        # Ex. {eax: 'eax'}
+        in_out.reject! { |k, v| k == v }
+
+        # Check input
+        if (in_out.keys - all_regs).any?
+          raise ArgumentError, format('Unknown register! Know: %p.  Got: %p', all_regs, in_out)
+        end
+
+        # Collapse constant values
+        #
+        # Ex. {eax: 1, ebx: 1} => {eax: 1, ebx: 'eax'}
+        # +post_mov+ are collapsed registers, set their values in the end.
+        post_mov = in_out.group_by { |_, v| v }.values.each_with_object({}) do |list, hash|
+          val = list.first[1]
+          next if list.size == 1 || all_regs.include?(val) || val.zero?
+          list.sort!
+          list[1..-1].each do |reg, _|
+            hash[reg] = list.first.first
+            in_out.delete(reg)
+          end
+        end
+
+        graph = in_out.dup
+        result = []
+
+        # Let's do the topological sort.
+        # so sad ruby 2.1 doesn't have +itself+...
+        deg = graph.values.group_by { |i| i }.map { |k, v| [k, v.size] }.to_h
+        graph.keys.each { |k| deg[k] = 0 unless deg.key?(k) }
+
+        until deg.empty?
+          min_deg = deg.min_by { |_, v| v }[1]
+          break unless min_deg.zero? # remain are all cycles
+          min_pivs = deg.select { |_, v| v == min_deg }
+          piv = randomize ? min_pivs.sample : min_pivs.first
+          dst = piv.first
+          deg.delete(dst)
+          next unless graph.key?(dst) # Reach an end node.
+          deg[graph[dst]] -= 1
+          result << ['mov', dst, graph[dst]]
+          graph.delete(dst)
+        end
+
+        # Remain must be cycles.
+        cycles = graph.keys.each_with_object([]) do |reg, obj|
+          next unless graph.key?(reg)
+          cycle = check_cycle(reg, graph)
+          obj << cycle
+          cycle.each { |r| graph.delete(r) }
+        end
+
+        cycles.each do |cycle|
+          # Try break a cycle
+          # 1. If +tmp+ is set, try to use it.
+          # 2. If +xchg == true+, use +xchg+.
+          # 3. Find proper +tmp+ and use it.
+          # 4. so sad :(.
+          if tmp && !depends_on_cycle(tmp, in_out, cycle) then result.concat(break_cycle(cycle, tmp: tmp))
```

Comment:
Seems i should not topological sort all simultaneously..
For example, `{a: 'b', b: 'a', c: 'a', x: 'y', y: 'x', z: 'x'}`
In this case, `x` can use for break `a <-> b` cycle, if I handle the `x y z` graph later than `a b c`.
Current implementation cannot handle this case even fix the issue you mentioned above.
I will add this test case to test and fix it later.